src/NHibernate.Test/Linq/FunctionTests.cs

```diff
@@ -114,17 +115,43 @@ public void LeftFunction()
 		[Test]
 		public void ReplaceFunction()
 		{
+			var suppliedName = "Anne";
 			var query = from e in db.Employees
 						where e.FirstName.StartsWith("An")
 						select new
-								{
-									Before = e.FirstName,
-									AfterMethod = e.FirstName.Replace("An", "Zan"),
-									AfterExtension = ExtensionMethods.Replace(e.FirstName, "An", "Zan"),
-									AfterExtension2 = e.FirstName.ReplaceExtension("An", "Zan")
-								};
-
-			var s = ObjectDumper.Write(query);
+							{
+								Before = e.FirstName,
+								// This one call the standard string.Replace, not the extension. The linq registry handles it.
+								AfterMethod = e.FirstName.Replace("An", "Zan"),
+								AfterExtension = ExtensionMethods.Replace(e.FirstName, "An", "Zan"),
+								AfterExtension2 = e.FirstName.ReplaceExtension("An", "Zan"),
+								AfterExtension3 = e.FirstName.ReplaceInDb("An", "Zan"),
+								BeforeConst = suppliedName,
+								// This one call the standard string.Replace, not the extension. The linq registry handles it.
+								AfterMethodConst = suppliedName.Replace("An", "Zan"),
+								AfterExtensionConst = ExtensionMethods.Replace(suppliedName, "An", "Zan"),
+								AfterExtension2Const = suppliedName.ReplaceExtension("An", "Zan"),
+								AfterExtension3Const = suppliedName.ReplaceInDb("An", "Zan"),
+							};
+			var results = query.ToList();
+			var s = ObjectDumper.Write(results);
+
+			foreach (var result in results)
+			{
+				var expectedDbResult = Regex.Replace(result.Before, "An", "Zan", RegexOptions.Compiled | RegexOptions.IgnoreCase);
+				Assert.That(result.AfterMethod, Is.EqualTo(expectedDbResult), "Wrong AfterMethod value");
+				Assert.That(result.AfterExtension, Is.EqualTo(expectedDbResult), "Wrong AfterExtension value");
+				Assert.That(result.AfterExtension2, Is.EqualTo(expectedDbResult), "Wrong AfterExtension2 value");
+				Assert.That(result.AfterExtension3, Is.EqualTo(expectedDbResult), "Wrong AfterExtension3 value");
+
+				var expectedDbResultConst = Regex.Replace(result.BeforeConst, "An", "Zan", RegexOptions.Compiled | RegexOptions.IgnoreCase);
+				var expectedInMemoryResultConst = result.BeforeConst.Replace("An", "Zan");
+				var expectedInMemoryExtensionResultConst = ExtensionMethods.Replace(result.BeforeConst, "An", "Zan");
+				Assert.That(result.AfterMethodConst, Is.EqualTo(expectedInMemoryResultConst), "Wrong AfterMethodConst value");
+				Assert.That(result.AfterExtensionConst, Is.EqualTo(expectedInMemoryExtensionResultConst), "Wrong AfterExtensionConst value");
+				Assert.That(result.AfterExtension2Const, Is.EqualTo(expectedInMemoryExtensionResultConst), "Wrong AfterExtension2Const value");
```

Comment:
When called on constant expressions, extension methods without `dbOnly = true` are called in-memory. This has always be the case with the NHibernate Linq provider.

(Since NH-3961 committed for v5, if the in-memory evaluation fails with an exception, the call stays in the expression instead of throwing. Even if the failure was not expected by the method developer.)